core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -267,376 +302,264 @@ protected void doRun() throws Exception {
+                logger.debug("[{}] creating index", index);
+                AllocatedIndex<? extends Shard> indexService = null;
+                try {
+                    indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, buildInIndexListener);
+                } catch (Throwable e) {
+                    sendFailShard(shardRouting, "failed to create index", e);
+                    failedIndices.add(index);
+                }
+                if (indexService != null) {
+                    updateIndexMapping(event, indexService, indexMetaData, failedIndices);
+                }
             }
-            indexService.updateMetaData(indexMetaData);
         }
     }
 
-    private void applyNewIndices(final ClusterChangedEvent event) {
-        // we only create indices for shards that are allocated
-        RoutingNode routingNode = event.state().getRoutingNodes().node(event.state().nodes().getLocalNodeId());
-        if (routingNode == null) {
+    private void updateIndices(ClusterChangedEvent event) {
+        if (!event.metaDataChanged()) {
             return;
         }
-        for (ShardRouting shard : routingNode) {
-            if (!indicesService.hasIndex(shard.index())) {
-                final IndexMetaData indexMetaData = event.state().metaData().getIndexSafe(shard.index());
-                if (logger.isDebugEnabled()) {
-                    logger.debug("[{}] creating index", indexMetaData.getIndex());
-                }
-                try {
-                    indicesService.createIndex(nodeServicesProvider, indexMetaData, buildInIndexListener);
-                } catch (Throwable e) {
-                    sendFailShard(shard, "failed to create index", e);
-                }
+        for (AllocatedIndex<? extends Shard> indexService : indicesService) {
+            Index index = indexService.index();
+            IndexMetaData indexMetaData = event.state().metaData().index(index);
+            assert indexMetaData != null : "index should have been deleted by applyDeleteIndices";
+            if (event.indexMetaDataChanged(indexMetaData)) {
+                indexService.updateMetaData(indexMetaData);
+                updateIndexMapping(event, indexService, indexMetaData, null);
             }
         }
     }
 
-    private void applyMappings(ClusterChangedEvent event) {
-        // go over and update mappings
-        for (IndexMetaData indexMetaData : event.state().metaData()) {
-            Index index = indexMetaData.getIndex();
-            if (!indicesService.hasIndex(index)) {
-                // we only create / update here
-                continue;
+    private void updateIndexMapping(ClusterChangedEvent event, AllocatedIndex<? extends Shard> indexService, IndexMetaData indexMetaData,
+                                    @Nullable Set<Index> failedIndices) {
+        try {
+            if (indexService.requiresIndexMappingRefresh(indexMetaData) && sendRefreshMapping) {
```

Comment:
I think `requiresIndexMappingRefresh` is very misleading - it should be called `updateMapping` as it does update the mapping (and returns a boolean for refreshes)